include/envoy/thread/thread.h

```diff
@@ -10,6 +10,16 @@
 namespace Envoy {
 namespace Thread {
 
+class ThreadId {
+public:
+  virtual ~ThreadId() {}
+
+  virtual std::string debugString() const PURE;
+  virtual bool isCurrentThreadId() const PURE;
+};
+
+typedef std::unique_ptr<ThreadId> ThreadIdPtr;
```

Comment:
I think we need to go with unique_ptr here (or shared_ptr) because the type can't be expressed concretely in the portable layer.  Is there a specific concern with unique_ptr here?

See http://man7.org/linux/man-pages/man3/pthread_self.3.html
```
POSIX.1 allows an implementation wide freedom in choosing the type used to
represent a thread ID; for example, representation using either an arithmetic
type or a structure is permitted. Therefore, variables of type pthread_t can't
portably be compared using the C equality operator (==); use pthread_equal(3)
instead.

Thread identifiers should be considered opaque: any attempt to use a
thread ID other than in pthreads calls is nonportable and can lead to
unspecified results.
```

Reply:
That's fine. I just would skip a level of indirection if not needed.